client-js/app/elements/labrad-plot.ts

```diff
@@ -470,206 +787,186 @@ export class Plot extends polymer.Base {
+    const yScreen = this.yScale(y) + (hScreen / 2);
 
-      default:
-        // Nothing to do.
-        break;
-    }
+    outputObject.x = xScreen;
+    outputObject.y = yScreen;
+    outputObject.w = wScreen;
+    outputObject.h = hScreen;
   }
 
 
-  private updatePlotStyles() {
-    if (!this.svg) return;
-    switch (this.numIndeps) {
-      case 1:
-        this.$$('rect.background').style.fill = '#ffffff';
-        this.$.modes2d.style.visibility = 'hidden';
-        this.is1D = true;
-        this.is2D = false;
-        break;
-
-      case 2:
-        this.$$('rect.background').style.fill = '#222222';
-        this.is1D = false;
-        this.is2D = true;
-        break;
-
-      default:
-        // Nothing to do.
-        break;
-    }
+  /**
+   * Projects a coordinate in screen coordinates (relative to the canvas) to
+   * world coordinates (relative to the 3D projection of the scene).
+   */
+  private projectScreenCoordToWorldCoord(x: number, y: number, outputVector: THREE.Vector3) {
+    outputVector.set((x / this.width) * 2 - 1, -(y / this.height) * 2 + 1, 0.5);
+    outputVector.unproject(this.camera);
   }
 
 
   /**
-   * Destroys and recreates the plot.
+   * Projects all graph data according to the latest zoom level.
+   *
+   * Due to the separation of each plot into a WebGL and SVG component, the
+   * zoom/pan of the d3 axes needs to lead to the updating of the WebGL
+   * display.
+   *
+   * Each object in the scene is iterated, its original graph data position is
+   * extracted and then reprojected to the correct position in WebGL (world)
+   * space.
+   *
+   * There are three primary coordinate systems:
+   * 1) Graph Space. The position of the data to be plotted in the graph.
+   * 2) Screen Space. The position relative to the plot on screen.
+   * 3) World Space. The position of the item in the 3D WebGL projected world.
    */
-  redraw(): void {
-    const area = this.$.plot,
-          rect = area.getBoundingClientRect();
-    while (area.firstChild) {
-      area.removeChild(area.firstChild);
+  private projectGraphPositions() {
+    // When plotting lines or single points, we only need one vertex to
+    // represent the data.
+    const numVertices = (this.numIndeps == 1 || this.drawMode2D == 'dots') ?
+        1 : this.planeVertexCount;
+
+    // A reusable buffer for manipulating the coordinates of each point.
+    const positionBuffer = new Float32Array(numVertices * 3);
+
+    // Reuse the same vector and object for all coordinate conversions to avoid
+    // generating a large amount of garbage causing GC stalls.
+    const vector = new THREE.Vector3();
+    const screenRect = { x: 0, y: 0, w: 0, h: 0 };
+
+    // Project the (0, 0) screen position to world coordinates so that widths
+    // and heights in world space can be calculated from the delta.
+    this.projectScreenCoordToWorldCoord(0, 0, vector);
+    const {
+      x: xWorldZero,
+      y: yWorldZero
+    } = vector;
```

Comment:
I think it only looks strange because destructuring is new syntax. It's ultimately a necessary style format that is required when you have longer destructurings. 

If you think of it as each being a variable assignment, this formatting makes it consistent with actual variable declarations, which are always done on their own lines.

```typescript
    const {
      x: xWorldZero,
      y: yWorldZero
    } = vector;
```

is the same as:

```typescript
    const xWorldZero = vector.x;
    const yWorldZero = vector.y;
```

Now we mention it, maybe these are all better off as just straight variable declarations.